caffe2/CMakeLists.txt

```diff
@@ -544,14 +545,15 @@ if (BUILD_PYTHON)
     set_target_properties(caffe2_pybind11_state_gpu PROPERTIES PREFIX "" DEBUG_POSTFIX "")
     set_target_properties(caffe2_pybind11_state_gpu PROPERTIES SUFFIX ${PY_EXT_SUFFIX})
     if (APPLE)
-      set_target_properties(caffe2_pybind11_state_gpu PROPERTIES LINK_FLAGS "-undefined dynamic_lookup")
+      set_target_properties(caffe2_pybind11_state_gpu PROPERTIES LINK_FLAGS "-undefined dynamic_lookup") 
     endif()
     target_include_directories(caffe2_pybind11_state_gpu PRIVATE $<INSTALL_INTERFACE:include>)
     target_include_directories(caffe2_pybind11_state_gpu PRIVATE ${Caffe2_CPU_INCLUDE})
     target_link_libraries(
         caffe2_pybind11_state_gpu caffe2_library caffe2_gpu_library)
     if (WIN32)
       target_link_libraries(caffe2_pybind11_state_gpu ${PYTHON_LIBRARIES})
+      target_link_libraries(caffe2_pybind11_state_gpu onnx_proto)
```

Comment:
Unfortunately, even after last explicit merge the error is here. We have to link to onnx_proto.

Beginning of error log to avoid spam:

> "D:\pytorch-scripts\caffe2_builders\v140\pytorch\build\Release\INSTALL.vcxproj" (целевой объект по умолчанию) (1) ->
>        "D:\pytorch-scripts\caffe2_builders\v140\pytorch\build\Release\ALL_BUILD.vcxproj" (целевой объект по умолчанию) (3) ->
>        "D:\pytorch-scripts\caffe2_builders\v140\pytorch\build\Release\caffe2\caffe2_pybind11_state.vcxproj" (целевой объект по умолчанию) (27) ->
>        (Целевой объект Link) ->
>          pybind_state.obj : error LNK2019: unresolved external symbol "public: __cdecl onnx_c2::ValueInfoProto::ValueInfoProto(void)" (??0ValueInfoProto@onnx_c2@@QEAA@XZ) referenced in func
>        tion "public: class std::vector<class std::vector<class pybind11::bytes,class std::allocator<class pybind11::bytes> >,class std::allocator<class std::vector<class pybind11::bytes,cla
>        ss std::allocator<class pybind11::bytes> > > > __cdecl <lambda_cbc5b9b4b7daf442305dec23ae9651fd>::operator()(class caffe2::onnx::Caffe2Backend &,class pybind11::bytes const &,class s
>        td::vector<class pybind11::bytes,class std::allocator<class pybind11::bytes> > const &,int)const " (??R<lambda_cbc5b9b4b7daf442305dec23ae9651fd>@@QEBA?AV?$vector@V?$vector@Vbytes@pyb
>        ind11@@V?$allocator@Vbytes@pybind11@@@std@@@std@@V?$allocator@V?$vector@Vbytes@pybind11@@V?$allocator@Vbytes@pybind11@@@std@@@std@@@2@@std@@AEAVCaffe2Backend@onnx@caffe2@@AEBVbytes@p
>        ybind11@@AEBV?$vector@Vbytes@pybind11@@V?$allocator@Vbytes@pybind11@@@std@@@2@H@Z) [D:\pytorch-scripts\caffe2_builders\v140\pytorch\build\Release\caffe2\caffe2_pybind11_state.vcxproj
>        ]

Reply:
BTW, the option is being explicitly set in new [msbuild.bat](https://github.com/peterjc123/pytorch-scripts/blob/master/caffe2_builders/v140/msbuild.bat) from @peterjc123's repo.
[auto.bat](https://github.com/peterjc123/pytorch-scripts/blob/master/caffe2_builders/v140/auto.bat) runs the whole build process by which I achieve this logs.

Reply:
Can you show me the full linking errors?

Reply:
You can reproduce it with the same script...

@peterjc123 

```
       "D:\pytorch-scripts\caffe2_builders\v140\pytorch\build\Debug\INSTALL.vcxproj" (целевой объект по умолчанию) (1) ->
       "D:\pytorch-scripts\caffe2_builders\v140\pytorch\build\Debug\ALL_BUILD.vcxproj" (целевой объект по умолчанию) (3) ->
       "D:\pytorch-scripts\caffe2_builders\v140\pytorch\build\Debug\caffe2\caffe2_pybind11_state.vcxproj" (целевой объект по умолчанию) (27) ->
       (Целевой объект Link) ->
         pybind_state.obj : error LNK2019: unresolved external symbol "public: __cdecl onnx_c2::ValueInfoProto::ValueInfoProto(void)" (??0ValueInfoProto@onnx_c2@@QEAA@XZ) referenced in func
       tion "public: class std::vector<class std::vector<class pybind11::bytes,class std::allocator<class pybind11::bytes> >,class std::allocator<class std::vector<class pybind11::bytes,cla
       ss std::allocator<class pybind11::bytes> > > > __cdecl <lambda_cbc5b9b4b7daf442305dec23ae9651fd>::operator()(class caffe2::onnx::Caffe2Backend &,class pybind11::bytes const &,class s
       td::vector<class pybind11::bytes,class std::allocator<class pybind11::bytes> > const &,int)const " (??R<lambda_cbc5b9b4b7daf442305dec23ae9651fd>@@QEBA?AV?$vector@V?$vector@Vbytes@pyb
       ind11@@V?$allocator@Vbytes@pybind11@@@std@@@std@@V?$allocator@V?$vector@Vbytes@pybind11@@V?$allocator@Vbytes@pybind11@@@std@@@std@@@2@@std@@AEAVCaffe2Backend@onnx@caffe2@@AEBVbytes@p
       ybind11@@AEBV?$vector@Vbytes@pybind11@@V?$allocator@Vbytes@pybind11@@@std@@@2@H@Z) [D:\pytorch-scripts\caffe2_builders\v140\pytorch\build\Debug\caffe2\caffe2_pybind11_state.vcxproj]
         pybind_state.obj : error LNK2019: unresolved external symbol "public: virtual __cdecl onnx_c2::ValueInfoProto::~ValueInfoProto(void)" (??1ValueInfoProto@onnx_c2@@UEAA@XZ) reference
       d in function "public: __cdecl std::pair<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const ,class onnx_c2::ValueInfoProto>::~pair<class st
       d::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const ,class onnx_c2::ValueInfoProto>(void)" (??1?$pair@$$CBV?$basic_string@DU?$char_traits@D@std@@V?$
       allocator@D@2@@std@@VValueInfoProto@onnx_c2@@@std@@QEAA@XZ) [D:\pytorch-scripts\caffe2_builders\v140\pytorch\build\Debug\caffe2\caffe2_pybind11_state.vcxproj]
         pybind_state.obj : error LNK2019: unresolved external symbol "public: __cdecl onnx_c2::ValueInfoProto::ValueInfoProto(class onnx_c2::ValueInfoProto &&)" (??0ValueInfoProto@onnx_c2@
       @QEAA@$$QEAV01@@Z) referenced in function "public: __cdecl std::pair<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const ,class onnx_c2::Val
       ueInfoProto>::pair<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const ,class onnx_c2::ValueInfoProto><class std::basic_string<char,struct s
       td::char_traits<char>,class std::allocator<char> >,class onnx_c2::ValueInfoProto,void,0>(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > &&,cl
       ass onnx_c2::ValueInfoProto &&)" (??$?0V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@VValueInfoProto@onnx_c2@@X$0A@@?$pair@$$CBV?$basic_string@DU?$char_traits@D@std@@
       V?$allocator@D@2@@std@@VValueInfoProto@onnx_c2@@@std@@QEAA@$$QEAV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@$$QEAVValueInfoProto@onnx_c2@@@Z) [D:\pytorch-scripts\caffe
       2_builders\v140\pytorch\build\Debug\caffe2\caffe2_pybind11_state.vcxproj]
         pybind_state.obj : error LNK2019: unresolved external symbol "public: __cdecl onnx_c2::TensorProto::TensorProto(void)" (??0TensorProto@onnx_c2@@QEAA@XZ) referenced in function "pub
       lic: class pybind11::bytes __cdecl <lambda_ec355006db76b188803f2103fd5aad3c>::operator()(class caffe2::onnx::Caffe2Backend &,class pybind11::bytes const &,class std::basic_string<cha
       r,struct std::char_traits<char>,class std::allocator<char> > const &)const " (??R<lambda_ec355006db76b188803f2103fd5aad3c>@@QEBA?AVbytes@pybind11@@AEAVCaffe2Backend@onnx@caffe2@@AEBV
       12@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z) [D:\pytorch-scripts\caffe2_builders\v140\pytorch\build\Debug\caffe2\caffe2_pybind11_state.vcxproj]
         pybind_state.obj : error LNK2019: unresolved external symbol "public: virtual __cdecl onnx_c2::TensorProto::~TensorProto(void)" (??1TensorProto@onnx_c2@@UEAA@XZ) referenced in func
       tion "public: class pybind11::bytes __cdecl <lambda_ec355006db76b188803f2103fd5aad3c>::operator()(class caffe2::onnx::Caffe2Backend &,class pybind11::bytes const &,class std::basic_s
       tring<char,struct std::char_traits<char>,class std::allocator<char> > const &)const " (??R<lambda_ec355006db76b188803f2103fd5aad3c>@@QEBA?AVbytes@pybind11@@AEAVCaffe2Backend@onnx@caf
       fe2@@AEBV12@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z) [D:\pytorch-scripts\caffe2_builders\v140\pytorch\build\Debug\caffe2\caffe2_pybind11_state.vcxproj]
         D:\pytorch-scripts\caffe2_builders\v140\pytorch\build\Debug\caffe2\python\Debug\caffe2_pybind11_state.cp35-win_amd64.pyd : fatal error LNK1120: 5 unresolved externals [D:\pytorch-s
       cripts\caffe2_builders\v140\pytorch\build\Debug\caffe2\caffe2_pybind11_state.vcxproj]
```

Reply:
@orionr @peterjc123 
BTW guys, while this is not yet approved I have an offtop quiestion.
Unfortunately, I'm not yet familiar who is responsible for which part of Caffe2.
My plan with Caffe2 is to use it as an Inference Engine for NVIDIA GPUs at my workplace. At the moment for Intel processors there is OpenVINO toolkit, and good contact with its developers.
Can you please tell me, who is responsible for verifying PRs of new ops, related to object detection?
I plan to implement [Normalize layer](https://github.com/weiliu89/caffe/blob/ssd/include/caffe/layers/normalize_layer.hpp) from ParseNet (for Caffe SSDs) and [CropAndResize](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/crop_and_resize_op.cc) from TF OD API. Also, there will be needed at least some modifications of the existing BoxWithNMSLimit/GenerateProposals or even implementation of separate [PriorBoxOp](https://github.com/weiliu89/caffe/blob/ssd/src/caffe/layers/prior_box_layer.cpp) and [DetectionOutputOp](https://github.com/weiliu89/caffe/blob/ssd/src/caffe/layers/detection_output_layer.cpp) to be able to run SSDs, trained in Caffe or TF OD API.
Who are people, who can review such changes?
This is sure an offtop, so I'll remove the question from here after answer. My e-mail is ArutyunovG@yandex.ru, if it is better to not answer in this thread.